core/src/util_normalize.cpp

```diff
@@ -81,36 +87,155 @@ bool normalize_nfd(std::u16string &str) {
+  if(!UASSERT_SUCCESS(status)) {
     return false;
   }
 
   dst.assign(udst.getBuffer(), udst.length());
   return true;
+#endif
+}
+
+bool
+normalize_nfd(km_core_usv cp, std::u32string &dst) {
+  // set the output string to the original string
+  dst.clear();
+  dst.append(1, cp);
+#ifdef __EMSCRIPTEN__
+  auto str16 = convert<char32_t, char16_t>(dst);
+  if (!normalize_nfd(str16)) {
+    return false;  // failed, retain original str
+  } else {
+    dst = convert<char16_t, char32_t>(str16);
+    return true;
+  }
+#else
+  UErrorCode icu_status = U_ZERO_ERROR;
+  const icu::Normalizer2 *nfd = icu::Normalizer2::getNFDInstance(icu_status);
+  assert(U_SUCCESS(icu_status));
+  if (!U_SUCCESS(icu_status)) {
+    // TODO: log the failure code
+    return false;
+  }
+  icu::UnicodeString decomposition;
+  if (!nfd->getDecomposition(cp, decomposition)) {
+    return false;  // no error, just no decomposition
+  } else {
+    dst.clear();
+    auto len = decomposition.countChar32();
+    for (int i = 0; i < len; i++) {
+      dst.append(1, decomposition.char32At(i));
+    }
+    return true;
+  }
+#endif
+}
+
+bool is_nfd(const std::u16string& str) {
+#ifdef __EMSCRIPTEN__
+  std::u16string o = str;
+  normalize_nfd(o);
+  return (o == str); // false if changed
+#else
+  UErrorCode status = U_ZERO_ERROR;
+  auto nfd = getNFD(status);
+  if (nfd == nullptr) return false;
+  auto ustr = icu::UnicodeString(false, str.c_str(), (int)str.length());
+  auto result = nfd->isNormalized(ustr, status);
+  if (!UASSERT_SUCCESS(status)) {
+    return false;
+  } else {
+    return result;
+  }
+#endif
+}
+
+bool is_nfd(const std::u32string& str) {
+#ifdef __EMSCRIPTEN__
+  std::u32string o = str;
+  normalize_nfd(o);
+  return (o == str); // false if changed
+#else
+  UErrorCode status = U_ZERO_ERROR;
+  auto nfd = getNFD(status);
+  if (nfd == nullptr) return false;
+  auto ustr = icu::UnicodeString::fromUTF32(reinterpret_cast<const UChar32*>(str.c_str()), (int)str.length());
+  auto result = nfd->isNormalized(ustr, status);
+  if (!UASSERT_SUCCESS(status)) {
+    return false;
+  } else {
+    return result;
+  }
+#endif
 }
 
+bool has_nfd_boundary_before(km_core_usv cp) {
+#ifdef __EMSCRIPTEN__
+// it's a negative table. entries in the table mean returning false. non-entries return true.
+  for (auto i=0;i<(km_noBoundaryBefore_entries*2);i+=2) {
+    auto start = km_noBoundaryBefore[i+0];
+    if (start > cp) return true;
+    auto count = km_noBoundaryBefore[i+1];
+    auto limit = start+count;
+    if (cp >= start && cp < limit) return false;
+  }
+  return true; // fallthrough
+#else
+  UErrorCode status = U_ZERO_ERROR;
+  auto nfd = getNFD(status);
+  if (nfd == nullptr) return false;
+  return nfd->hasBoundaryBefore(cp);
+#endif
+}
+
+/**
+ * Helper to convert icu::UnicodeString to a UTF-32 km_core_usv buffer,
+ * nul-terminated
+ */
+km_core_usv *string_to_usv(const std::u32string& src) {
+  return km::core::kmx::u32dup(src.c_str());
+}
```

Comment:
Uh... so `std::u32string` doesn't look the same as `icu::UnicodeString` to me.  Is the comment saying that we're basically just using the type to store `icu::UnicodeString` values without the need to link in its type for WASM, with this method doing the _actual_ conversion?

Without the comment, my naive sight-reading of the code suggests that this is simply cloning the string.  It looks like `km_core_usv` is almost just an alias for `std::u32string`, but something tells me that's probably not _entirely_ right.

Reply:
The comment is wrong here… If this function is retained, it's just a way to make the ICU and non-ICU versions a little bit more parallel

Reply:
I'll have to take your word for it - I don't have sufficient context to review details related to your comment.